### PR TITLE
[SOFT-465] Separate logging in MU

### PR DESCRIPTION
**Pull request by jessm**
idk it's kinda messy, comments appreciated

This is so in future we can easily
- toggle logging for pm
- assert on logs maybe??
- serve logs via muctl

### CODE REVIEW

**Review by copperium — CHANGES_REQUESTED** [review]
Looks good, only thing blocking is the `log.tags` typo. Love the pub/sub stuff.

File: mu/harness/logger.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        for sub in self.subs.values():
+            sub.put(l)
+
+    def query(self, tag):
+        ret = []
+        for log in self.logs:
+            if log.tag == tag:
+                ret.append(log)
+
+        return ret
```
[the remark this answers is not in the document]
Comment (jessm) [reply mu/harness/logger.py]: true

File: mu/harness/logger.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
+from collections import deque
+import queue
+
+class NoLog(queue.Empty):
+    pass
+
+class Log:
+    def __init__(self, tag, msg):
+        self.tag = tag
+        self.msg = msg
```
[the remark this answers is not in the document]
Comment (jessm) [reply mu/harness/logger.py]: true

File: mu/harness/logger.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+class Subscriber:
+    # Log consumers create a subscriber and add it to the logger
+    # tags is an iterable with the tags to subscribe to
+    def __init__(self, name, tags=None):
+        self.name = name
+        self.tags = tags
+        self.q = queue.Queue()
+
+    def put(self, log):
+        if not self.tags or log.tags in self.tags:
```
[the remark this answers is not in the document]
Comment (jessm) [reply mu/harness/logger.py]: true

**Review by copperium — APPROVED** [review]

**Pull request merged**